src/runtime/HalideBuffer.h

```diff
@@ -34,6 +34,28 @@
 #pragma GCC diagnostic ignored "-Warray-bounds"
 #endif
 
+#if defined(HALIDE_RUNTIME_BUFFER_WRAPPERS)
+
+template<typename T>
+struct halide_typed_buffer_t {
+  HALIDE_ALWAYS_INLINE explicit halide_typed_buffer_t(halide_buffer_t *p) : ptr(p) {
+    using not_const_T = typename std::remove_const<T>::type;
+    constexpr bool T_is_void = std::is_same<not_const_T, void>::value;
+    static_assert(!T_is_void, "Cannot create halide_typed_buffer_t<void>");
+    using not_void_T = typename std::conditional<T_is_void, uint8_t, not_const_T>::type;
+    assert(p->type == halide_type_of<not_void_T>());
+  }
+
+  halide_buffer_t * const ptr;
+};
+
+template<typename T>
+HALIDE_ALWAYS_INLINE
+halide_buffer_t *halide_typed_buffer_t_to_halide_buffer_t(const halide_typed_buffer_t<T> &t) { return t.ptr; }
```

Comment:
What are the advantages of this intermediate type over just a free function that takes whatever Buffer type the user has provided?

Reply:
If we have a free function, what will the type of the argument be?

- If it's `BufType&`, the arg has to be an lvalue, so you can't pass temporaries.
- If it's `const BufType&`, then we have to do a `const_cast` internally, and I'm convinced that every time you do a `const_cast`, God kills a kitten.
- if it's just plain `BufType`, you may end up with unnecessary copies (if the value isn't provably movable by the compiler).

Reply:
I tried `BufType &&` but I can't get the return type to work. A temporary object seems to be necessary.

Reply:
There may well be a slicker way to do it, but until/unless we find someone who really needs the extra flexibility, this works.